kahuna/public/js/services/api/leases.js

```diff
@@ -5,87 +5,100 @@ import '../../services/image-list';
+            return refreshImages([image]).then(images => {
+                const currentLeases = getLeases(images);
+                return currentLeases.then(() => {
+                    return image
+                        .perform('delete-leases').then(() => {
+                            pollLeasesAndUpdateUI(images, imageList.getLeases(images));
+                        });
+                });
+            });
+        }
 
-        return currentLeases.then(() => {
+        function replace(image, leases) {
+            const images = [image];
+            const currentLeasesPromise = getLeases(images);
 
-            const updatedLeases = leases.map((lease) => {
-                var newLease = angular.copy(lease);
-                newLease.mediaId = image.data.id;
-                return newLease;
-            });
+            return currentLeasesPromise.then((currentLeases) => {
 
-            return image
-              .perform('replace-leases', {body: updatedLeases})
-              .then(() => {
-                  pollLeasesAndUpdateUI(images);
-              });
-        });
-    }
+                const updatedLeases = leases.map((lease) => {
+                    var newLease = angular.copy(lease);
+                    newLease.mediaId = image.data.id;
+                    return newLease;
+                });
 
-    function add(image, lease) {
-      const newLease = angular.copy(lease);
-      newLease.mediaId = image.data.id;
+                // Don't update the leases if they're "the same"
+                if (JSON.stringify(currentLeases.leases) === JSON.stringify(updatedLeases)) {
+                    return image;
+                }
 
-      if (angular.isDefined(newLease.notes) && newLease.notes.trim().length === 0) {
-        newLease.notes = null;
-      }
+                return image
+                    .perform('replace-leases', { body: updatedLeases })
+                    .then(() => pollLeasesAndUpdateUI(images));
+            });
+        }
 
-      return image.perform('add-lease', {body: newLease});
-    }
+        function add(image, lease) {
+            const newLease = angular.copy(lease);
+            newLease.mediaId = image.data.id;
 
-    function batchAdd(lease, images) {
-      return trackAll($rootScope, "leases", images, [image => add(image, lease), image => {
-        apiPoll(() => {
-          return untilLeasesChange([image]);
-        });
-      }], 'images-changed').then(() => {
-            $rootScope.$emit('leases-updated');
-      });
+            if (angular.isDefined(newLease.notes) && newLease.notes.trim().length === 0) {
+                newLease.notes = null;
+            }
+            return image.perform('add-lease', { body: newLease });
+        }
+
+        function batchAdd(lease, images) {
+            //We check whether the leases in the image have a later lastModified date,
+            //so make sure that we don't have anything in the pipeline.
```

Comment:
Re: `make sure that we don't have anything in the pipeline.`, unsure how this maps to the logic below (what's making sure of this?)

Reply:
This relates more to `refreshImages(images).then` will move.

Reply:
Moved.